Source/objects.cpp

```diff
@@ -2512,20 +2512,23 @@ void OperateShrineEnchanted(Player &player)
 	if (cnt > 1) {
 		spell = 1;
 		for (int j = static_cast<int8_t>(SpellID::Firebolt); j < maxSpells; j++) { // BUGFIX: < MAX_SPELLS, there is no spell with MAX_SPELLS index (fixed)
-			if ((player._pMemSpells & spell) != 0) {
-				if (player._pSplLvl[j] < MaxSpellLevel)
-					player._pSplLvl[j]++;
+			if ((player._pMemSpells & spell) != 0 && player._pSplLvl[j] < MaxSpellLevel) {
+				NetSendCmdParam2(true, CMD_CHANGE_SPELL_LEVEL, static_cast<uint16_t>(j), static_cast<uint8_t>(player._pSplLvl[j] + 1));

```

Comment:
Should we maybe set this one to low pri, so that it is not sent before all of them are changed, and won't risk a race condition where the one that lowers the level arrives before the one that increases it?

```suggestion
				NetSendCmdParam2(false, CMD_CHANGE_SPELL_LEVEL, static_cast<uint16_t>(j), static_cast<uint8_t>(player._pSplLvl[j] + 1));
```

Reply:
I updated the code to make it more straight forward and only update every spell only once.
Also it fixes the bug that a maxed spell is reduced by 2 instead of 1.

Reply:
If `gbShouldValidatePackage` is false when sending a high-priority packet, the game creates a new packet to be sent out to all other players. It copies as much data as it can from the high-priority buffer, then the low-priority buffer, and finally fills the rest with monster sync data. Anything from the high-priority or low-priority buffers that couldn't fit into the packet will stay behind to be sent with the next packet. This means that high-priority packets could be sent before low-priority packets regardless of the order in which `NetSendCmd*` functions were called, but all packets with the same priority as each other should be sent in the same order as the function calls.

Reply:
It occurs to me after explaining this that it wouldn't just create a race condition. The packet sent on line 2531 would _always_ be sent before the ones sent on this line because these would only be written to the low-priority buffer. At the earliest, they would be sent to remote clients after 2531 writes its packet to the high-priority buffer.

Reply:
Thanks for explaining that.